sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -891,6 +1014,50 @@ public void setValueDeserializer(String valueDeserializer) {
       return input.getPipeline().apply(transform);
     }
 
+    private Coder<K> getKeyCoder(CoderRegistry coderRegistry, String schemaRegistryURL) {
+      Schema avroKeySchema = null;
+      if (schemaRegistryURL != null && getKeySchemaSubject() != null) {
+        avroKeySchema = fetchAvroSchema(schemaRegistryURL, getKeySchemaSubject());
+        checkArgument(avroKeySchema != null, "Avro key schema can't be null");
+      }
+
+      Coder<K> keyCoder;
+      if (avroKeySchema != null) {
+        keyCoder = (Coder<K>) AvroCoder.of(avroKeySchema);
+      } else {
+        keyCoder =
+            getKeyCoder() != null ? getKeyCoder() : inferCoder(coderRegistry, getKeyDeserializer());
+      }
+      checkState(
+          keyCoder != null,
+          "Key coder could not be inferred from key deserializer. Please provide"
+              + "key coder explicitly using withKeyDeserializerAndCoder()");
+      return keyCoder;
+    }
+
+    private Coder<V> getValueCoder(CoderRegistry coderRegistry, String schemaRegistryURL) {
+      Schema avroValueSchema = null;
+      if (schemaRegistryURL != null && getValueSchemaSubject() != null) {
+        avroValueSchema = fetchAvroSchema(schemaRegistryURL, getValueSchemaSubject());
+        checkArgument(avroValueSchema != null, "Avro value schema can't be null");
+      }
+
+      Coder<V> valueCoder;
+      if (avroValueSchema != null) {
+        valueCoder = (Coder<V>) AvroCoder.of(avroValueSchema);
+      } else {
+        valueCoder =
+            getValueCoder() != null
+                ? getValueCoder()
+                : inferCoder(coderRegistry, getValueDeserializer());
+      }
+      checkState(
+          valueCoder != null,
+          "Value coder could not be inferred from value deserializer. Please provide"
+              + "value coder explicitly using withValueDeserializerAndCoder()");
+      return valueCoder;
+    }
```

Comment:
It would be very useful to set a Beam schema based on the Avro schema instead of using AvroCoder. This would allow KafkaIO to be used as a source for Beam SQL

Reply:
Excellent idea, I would suggest to use a more conservative approach to enable Beam Schema Coders similar to the one used for `set/getInferBeamSchema` in `AvroIO`.

Reply:
I planned to do this in the next PR for https://issues.apache.org/jira/browse/BEAM-7336